tests/load/orchestrator.go

```diff
@@ -0,0 +1,315 @@
+		}
+
+		if issuerCtx.Err() != nil {
+			break // Case 1
+		}
+
+		currConfirmed := o.tracker.GetObservedConfirmed()
+		currTime := time.Now()
+
+		if o.config.MaxTPS == -1 {
+			return currTargetTPS.Load() == int64(currConfirmed)
+		}
+
+		tps := computeTPS(prevConfirmed, currConfirmed, currTime.Sub(prevTime))
+		o.setMaxObservedTPS(int64(tps))
+
+		// if max TPS target has been reached and we don't terminate, then continue here
+		// so we do not keep increasing the target TPS
+		if achievedTargetTPS && !o.config.Terminate {
+			o.log.Info(
+				"current network state",
+				zap.Uint64("current TPS", tps),
+				zap.Int64("max observed TPS", o.maxObservedTPS.Load()),
+			)
+			continue
+		}
+
+		if int64(tps) >= currTargetTPS.Load() {
+			if currTargetTPS.Load() >= o.config.MaxTPS {
+				achievedTargetTPS = true
+				o.log.Info(
+					"max TPS target reached",
+					zap.Int64("max TPS target", currTargetTPS.Load()),
+					zap.Uint64("average TPS", tps),
+				)
+				if o.config.Terminate {
+					o.log.Info("terminating orchestrator")
+					break // Case 2
+				}
+				o.log.Info("orchestrator will now continue running at max TPS")
+				continue
+			}
+			o.log.Info(
+				"increasing TPS",
+				zap.Int64("previous target TPS", currTargetTPS.Load()),
+				zap.Uint64("average TPS", tps),
+				zap.Int64("new target TPS", currTargetTPS.Load()+o.config.Step),
+			)
+			currTargetTPS.Add(o.config.Step)
+			attempts = 1
+		} else {
+			if attempts >= o.config.MaxAttempts {
+				o.log.Info(
+					"max attempts reached",
+					zap.Int64("attempted target TPS", currTargetTPS.Load()),
+					zap.Uint64("number of attempts", attempts),
+				)
+				break // Case 3
+			}
+			o.log.Info(
+				"failed to reach target TPS, retrying",
+				zap.Int64("current target TPS", currTargetTPS.Load()),
+				zap.Uint64("average TPS", tps),
+				zap.Uint64("attempt number", attempts),
+			)
+			attempts++
+		}
+
+		prevConfirmed = currConfirmed
+		prevTime = currTime
+	}
+
+	return achievedTargetTPS
+}
+
+// GetObservedIssued returns the max TPS the orchestrator observed
+func (o *Orchestrator[_]) GetMaxObservedTPS() int64 {
+	return o.maxObservedTPS.Load()
+}
+
+// start a goroutine to each issuer to continuously send transactions.
+// if an issuer errors, all other issuers will stop as well.
+func (o *Orchestrator[_]) issueTxs(ctx context.Context, currTargetTPS *atomic.Int64) {
+	for _, agent := range o.agents {
+		o.issuerGroup.Go(func() error {
+			for {
+				if ctx.Err() != nil {
+					return nil //nolint:nilerr
+				}
+				currTime := time.Now()
+				txsPerIssuer := uint64(math.Ceil(float64(currTargetTPS.Load())/float64(len(o.agents))) * o.config.TxRateMultiplier)
+				// always listen until listener context is cancelled, do not call agent.Listener.IssuingDone().
+				for range txsPerIssuer {
+					tx, err := agent.Issuer.GenerateAndIssueTx(ctx)
+					if err != nil {
+						return err
+					}
+					agent.Listener.RegisterIssued(tx)
```

Comment:
the orchestrator could call o.tracker.Issue(tx) instead of the issuer calling the tracker